### PR TITLE
Run Integration tests on DevOps Pipeline

### PR DESCRIPTION
**Pull request by avagraw**
---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] If introducing new functionality or modified behavior, are they backed by unit and integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** unit **and** integration tests passed locally? i.e. `pytest <project root> -vv`
- [ ] Have static checks passed using the .pylintrc and .flake8 rules? Look at the CI scripts for example usage.
- [ ] Have you made an entry in HISTORY.rst which concisely explains your feature or change?

### CODE REVIEW

File: azext_iot/common/deps.py [hunk 1]
```diff
@@ -12,11 +12,19 @@
...
 from azext_iot.common._homebrew_patch import HomebrewPipPatch
+from azure.cli.core._profile import Profile
+from azure.cli.core.mock import DummyCli
+from azext_iot.tests.settings import UserTypes
 
 
 def ensure_uamqp(config, yes=False, repair=False):
     if get_uamqp_ext_version(config) != EVENT_LIB[1] or repair or not test_import(EVENT_LIB[0]):
-        if not yes:
+        profile = Profile(cli_ctx=DummyCli())
```
Comment (digimaun) [inline azext_iot/common/deps.py:22]: Please don't use `DummyCli` in non-test code.  Also why is this needed, providing a `-y` to a uamqp command should skip the interaction flow?

File: azext_iot/tests/__init__.py [hunk 2]
```diff
@@ -126,16 +127,29 @@ def __init__(self, test_scenario):
...
                 ).get_output_in_json()
 
+                profile = Profile(cli_ctx=DummyCli())
+                subscription = profile.get_subscription()
+                user = subscription["user"]
+
+                if user["type"] == UserTypes.user.value:
+                    user_name = user["name"]
+                elif user["type"] == UserTypes.servicePrincipal.value:
+                    if settings.env.azext_iot_testserviceprincipal:
```
Comment (digimaun) [inline azext_iot/tests/__init__.py:141]: What is the purpose of requiring a new test variable `azext_iot_testserviceprincipal`? That can be eliminated -- `az account show`  will show the currently logged in principal and type and `az ad [user|sp] show` will provide lots of additional details such as objectId.

And the `EmbeddedCLI` can be used to call other commands - there is prior art for this.

File: .azure-devops/templates/nightly-tests.yml [hunk 3]
```diff
@@ -32,12 +32,12 @@ steps:
...
     displayName: 'All integration tests'
     inputs:
-      azureSubscription: az-cli-nightly
+      azureSubscription: az_cli_nightly_conn
       scriptType: bash
       scriptLocation: inlineScript
       inlineScript: |
         export COVERAGE_FILE=.coverage.all
-        pytest -vv azext_iot/tests -k "_int" --cov=azext_iot --cov-config .coveragerc --junitxml=junit/test-iotext-int.xml
+        pytest -vv azext_iot/tests/iothub -k "_int" --dist=loadfile -n 12 --reruns 2 --reruns-delay 60 --cov=azext_iot --cov-config .coveragerc --junitxml=junit/test-iotext-int.xml
```
Comment (digimaun) [inline .azure-devops/templates/nightly-tests.yml:40]: This is constraining pytest to only run iothub tests rather than all tests across different areas like before.

Comment (digimaun) [inline .azure-devops/templates/nightly-tests.yml:40]: Parallelization threshold should be configurable, not hard-coded.

Comment (avagraw) [reply .azure-devops/templates/nightly-tests.yml:40]: Yeah this will be addressed as soon as I able to get the permission issues for central and dps tests

Comment (avagraw) [reply .azure-devops/templates/nightly-tests.yml:40]: Yeah I am testing with the optimal value for number of hubs, will make this configurable and default to the optimal value

File: azext_iot/tests/__init__.py (hunk 2 above)
Comment (avagraw) [reply azext_iot/tests/__init__.py]: Thanks the logic got super simplified with just ```az account show```. The account show gives the service principal's app id which can be used as assignee for role assignment

**Comment by avagraw** [comment]
Testing evidence: https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=3202&view=logs&s=859b8d9a-8fd6-5a5c-6f5e-f84f1990894e&j=99bc3f8f-fb7d-5b02-d266-c92c3a07560b

**Comment by avagraw** [comment]
Release pipeline testing evidence: https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=3244&view=logs&j=bf6c90a3-cd0e-5c60-baf1-5d9ad2af1063&t=c6c5483d-c67b-5b91-bbbd-84cb21d06343

After the parallelization we have been able to achieve the following improvements in test execution times:

1. IoT Hub tests now finish in ~15 mins, they used to take ~35 mins earlier.
2. Central tests finish in ~10 mins, they used to take ~23 mins earlier
3. Digital Twins test now finish in ~8 mins, they used to take ~20 mins earlier

File: .azure-devops/templates/nightly-tests.yml [hunk 4]
```diff
@@ -30,15 +39,48 @@ steps:
...
 
+  - task: AzureCLI@2
+    continueOnError: true
+    displayName: 'IoTHub integration tests - jobs, hub and configuration'
+    inputs:
+      azureSubscription: az_cli_nightly_conn
+      scriptType: bash
+      scriptLocation: inlineScript
+      inlineScript: |
+        export COVERAGE_FILE=.coverage.all
```
Comment (digimaun) [inline .azure-devops/templates/nightly-tests.yml:59]: Doesn't look like code coverage is being calculated right for nightly, because the coverage file has the same name - the last set of tests that run will be included in code coverage.

File: .azure-devops/create-release.yml [hunk 5]
```diff
@@ -97,24 +97,33 @@ stages:
...
+    - job: 'testHub_job_1'
+      displayName: 'Test IoT Hub - config, core and jobs'
       condition: eq('${{ parameters.testHub }}', true)
       steps:
-      - template: templates/run-tests.yml
+      - template: templates/run-tests-parallel.yml
         parameters:
-          path: 'azext_iot/tests/iothub'
+          path: 'azext_iot/tests/iothub/configurations azext_iot/tests/iothub/core azext_iot/tests/iothub/jobs'
+          name: 'iot-hub'
```
Comment (digimaun) [inline .azure-devops/create-release.yml:112]: This needs a different name to not conflict with distributed code coverage combining

File: .azure-devops/templates/run-tests-parallel.yml [hunk 6]
```diff
@@ -0,0 +1,99 @@
+parameters:
```
Comment (digimaun) [inline .azure-devops/templates/run-tests-parallel.yml:1]: Can nightly use `run-tests-parallel`?

File: azext_iot/tests/__init__.py [hunk 7]
```diff
@@ -28,14 +31,23 @@
 USER_ROLE = "IoT Hub Data Contributor"
 DEFAULT_CONTAINER = "devices"
 
+APP_ID = os.environ.get("azext_iot_central_app_id")
```
Comment (digimaun) [inline azext_iot/tests/__init__.py:34]: Central test elements should be placed in `tests/central/[__init__].py` rather than the root tests init

File: azext_iot/tests/__init__.py [hunk 8]
```diff
@@ -86,6 +98,210 @@ def command_execute_assert(self, command, asserts):
 
         return output
 
+    def _create_device(self, **kwargs) -> (str, str):
```
Comment (digimaun) [inline azext_iot/tests/__init__.py:101]: Continuing on [this comment](https://github.com/Azure/azure-iot-cli-extension/pull/403/files#r687213101), Central test helpers should be organized under `tests/central/*` like how digitaltwins tests and related parts are organized.

Comment (digimaun) [reply azext_iot/tests/__init__.py:101]: `CaptureOutputLiveScenarioTest` is meant to be a generic test class for capturing output, it is not meant to encapsulate iot central test helpers.

File: azext_iot/tests/iothub/devices/test_iothub_nested_edge_int.py [hunk 9]
```diff
@@ -195,6 +196,9 @@ def test_iothub_nested_edge(self):
                 checks=self.is_empty(),
             )
 
+            # Wait for child devices to be removed to prevent failures
+            sleep(60)
```
Comment (digimaun) [inline azext_iot/tests/iothub/devices/test_iothub_nested_edge_int.py:200]: 60 seconds for this seems a long time, does shortening it to 15-30 seconds have failures?

File: azext_iot/tests/iothub/core/test_iot_ext_int.py [hunk 10]
```diff
@@ -7,15 +7,18 @@
 import os
 import pytest
 from time import sleep
+from knack.util import CLIError
```
Comment (digimaun) [inline azext_iot/tests/iothub/core/test_iot_ext_int.py:10]: The module `test_iot_ext_int.py` has undergone a lot of changes and removals of prior tests in the last few months, we should change the name to something that better reflects its test focus.

File: azext_iot/tests/iothub/core/test_iot_ext_int.py [hunk 11]
```diff
@@ -101,7 +130,7 @@ def tearDown(self):
     def test_storage(self):
         device_count = 1
 
-        content_path = os.path.join(CWD, "test_generic_replace.json")
+        content_path = os.path.join(CWD, "../test_generic_replace.json")
```
Comment (digimaun) [inline azext_iot/tests/iothub/core/test_iot_ext_int.py:133]: Instead of the `../`, something like this is more env agnostic `os.path.join(os.path.dirname(CWD), "test_generic_replace.json")`

Comment (digimaun) [inline azext_iot/tests/iothub/core/test_iot_ext_int.py:133]: Instead of the `../`, something like this is more env agnostic `os.path.join(os.path.dirname(CWD), "test_generic_replace.json")`

File: .azure-devops/templates/run-tests-parallel.yml [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+            pytest -vv ${{ parameters.path }} -k "_int.py" --cov=azext_iot --cov-config .coveragerc --junitxml=junit/test-iotext-int.xml
+        ${{ if containsValue(parameters.parallel_execution_dirs, parameters.path) }}:
+          inlineScript: |
+            export COVERAGE_FILE=.coverage.${{ parameters.name }}
+            pytest -vv ${{ parameters.path }} -k "_int.py" --dist=loadfile -n ${{ parameters.num_threads }} --reruns ${{ parameters.num_reruns }} --reruns-delay ${{ parameters.reruns_delay }} --cov=azext_iot --cov-config .coveragerc --junitxml=junit/test-iotext-int.xml        
+        ${{ if contains(parameters.path, 'azext_iot/tests/iothub') }}:
+          inlineScript: |
+            export COVERAGE_FILE=.coverage.${{ parameters.name }}
+            pytest -vv ${{ parameters.path }} -k "_int.py" --dist=loadfile -n ${{ parameters.num_iothub_threads }} --reruns ${{ parameters.num_reruns }} --reruns-delay ${{ parameters.reruns_delay }} --cov=azext_iot --cov-config .coveragerc --junitxml=junit/test-iotext-int.xml --deselect=azext_iot/tests/iothub/core/test_iothub_discovery_int.py::TestIoTHubDiscovery::test_iothub_targets
```
Comment (digimaun) [inline .azure-devops/templates/run-tests-parallel.yml:84]: Good use of `--deselect` as workaround for test issue

File: .azure-devops/templates/run-tests-parallel.yml (hunk 6 above)
Comment (digimaun) [reply .azure-devops/templates/run-tests-parallel.yml:1]: of course parameterizing what needs to be like the service connection

File: azext_iot/tests/iothub/devices/test_iothub_nested_edge_int.py (hunk 9 above)
Comment (avagraw) [reply azext_iot/tests/iothub/devices/test_iothub_nested_edge_int.py:200]: yeah 30 seconds was causing intermittent failures.

**Comment by avagraw** [comment]
Testing evidence:

Nightly pipeline: https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=3325&view=logs&s=1e1b1bb4-dc40-5f24-6e3b-50930d2876f6

The nightly pipeline was able to finish all tests in ~13 mins

Release pipeline: https://dev.azure.com/azureiotdevxp/aziotcli/_build/results?buildId=3331&view=logs&j=bf6c90a3-cd0e-5c60-baf1-5d9ad2af1063

**Review by digimaun — APPROVED** [review]
